agent/grpc-external/services/resource/read.go

```diff
@@ -60,7 +60,7 @@ func (s *Server) Read(ctx context.Context, req *pbresource.ReadRequest) (*pbreso
 	}
 
 	// Check tenancy exists for the V2 resource.
-	if err = tenancyExists(reg, s.TenancyBridge, req.Id.Tenancy, codes.NotFound); err != nil {
+	if err = tenancyExists(reg, s.TenancyBridge, req.Id.Tenancy, codes.InvalidArgument); err != nil {
```

Comment:
`codes.NotFound` is actually correct here. I verified against the v1 api:

```
$ consul partition create -name ap1
$ consul namespace create -name ns1 -partition ap1
$ consul kv put -partition ap1 -namespace ns1 key1 value1
$ curl http://localhost:8500/v1/kv/key1\?partition\=ap1\;namespace\=ns1
[
    {
        "LockIndex": 0,
        "Key": "key1",
        "Flags": 0,
        "Value": "dmFsdWUx",
        "Partition": "ap1",
        "Namespace": "ns1",
        "CreateIndex": 28,
        "ModifyIndex": 28
    }
]

# ns2 doesn't exist
$ curl -vvv http://localhost:8500/v1/kv/key1\?partition\=ap1\&namespace\=ns2

*   Trying 127.0.0.1:8500...
* Connected to localhost (127.0.0.1) port 8500 (#0)
> GET /v1/kv/key1?partition=ap1&namespace=ns2 HTTP/1.1
> Host: localhost:8500
> User-Agent: curl/8.1.2
> Accept: */*
>
< HTTP/1.1 404 Not Found
< Vary: Accept-Encoding
< X-Consul-Default-Acl-Policy: allow
< X-Consul-Index: 1
< X-Consul-Knownleader: true
< X-Consul-Lastcontact: 0
< X-Consul-Query-Backend: blocking-query
< Date: Tue, 24 Oct 2023 18:24:40 GMT
< Content-Length: 0
<
```

Reply:
Hmm, I see. In the other places(listbyowner, write_status, and write)  where `tenancyExists` call happens, we return `InvalidArgument`. We perhaps need those changed then.

Reply:
I would say each endpoint should be evaluated in isolation based on the expectations for that endpoint. While it may sound correct for it to be consistent across all endpoints, ultimately it should be an expected response for the operation being performed.

e.g. If you perform a `read`, you would expect a not found error because the thing you're attempting to read is something that doesn't exist. If you're performing a `write`, invalid argument makes sense because you're attempting to mutate the system with an invalid value. If you're performing a `list`, the result is just an empty array with a success response.
 
Additionally, it would be useful to observe the behavior with the v1 counterparts to factor the existing behavior into the equation (since we're maintaining feature parity).

Reply:
Alright, I will stick with the `NotFound` for read given that the `v1` does the same.